rust_snuba/src/strategies/clickhouse/writer_v2.rs

```diff
@@ -0,0 +1,215 @@
+        let http_port = cluster_config.http_port;
+        let database = cluster_config.database;
+
+        let inner = RunTaskInThreads::new(
+            next_step,
+            clickhouse_task_runner(
+                Arc::new(ClickhouseClient::new(
+                    &hostname, http_port, &table, &database,
+                )),
+                skip_write,
+            ),
+            concurrency,
+            Some("clickhouse"),
+        );
+
+        ClickhouseWriterStep { inner }
+    }
+}
+
+impl<N> ProcessingStrategy<BytesInsertBatch<RowData>> for ClickhouseWriterStep<N>
+where
+    N: ProcessingStrategy<BytesInsertBatch<()>>,
+{
+    fn poll(&mut self) -> Result<Option<CommitRequest>, StrategyError> {
+        self.inner.poll()
+    }
+
+    fn submit(
+        &mut self,
+        message: Message<BytesInsertBatch<RowData>>,
+    ) -> Result<(), SubmitError<BytesInsertBatch<RowData>>> {
+        self.inner.submit(message)
+    }
+
+    fn terminate(&mut self) {
+        self.inner.terminate();
+    }
+
+    fn join(&mut self, timeout: Option<Duration>) -> Result<Option<CommitRequest>, StrategyError> {
+        self.inner.join(timeout)
+    }
+}
+
+#[derive(Clone)]
+pub struct ClickhouseClient {
+    client: Client,
+    headers: HeaderMap<HeaderValue>,
+    url: String,
+    query: String,
+}
+
+impl ClickhouseClient {
+    pub fn new(hostname: &str, http_port: u16, table: &str, database: &str) -> ClickhouseClient {
+        let mut headers = HeaderMap::with_capacity(3);
+        headers.insert(CONNECTION, HeaderValue::from_static("keep-alive"));
+        headers.insert(ACCEPT_ENCODING, HeaderValue::from_static("gzip,deflate"));
+        headers.insert(
+            "X-ClickHouse-Database",
+            HeaderValue::from_str(database).unwrap(),
+        );
+
+        let query_params = "load_balancing=in_order&insert_distributed_sync=1".to_string();
+        let url = format!("http://{hostname}:{http_port}?{query_params}");
+        let query = format!("INSERT INTO {table} FORMAT JSONEachRow");
+
+        ClickhouseClient {
+            client: Client::new(),
+            headers,
+            url,
+            query,
+        }
+    }
+
+    fn chunked_body(rows: &[u8]) -> Vec<Result<Box<[u8]>, std::io::Error>> {
+        rows.chunks(CLICKHOUSE_HTTP_CHUNK_SIZE)
+            .map(|chunk| Ok(Box::from(chunk)))
+            .collect()
+    }
+
+    pub async fn send(&self, body: &[u8]) -> anyhow::Result<Response> {
+        let chunks = Self::chunked_body(body);
+        let stream = futures::stream::iter(chunks);
+
+        let res = self
+            .client
+            .post(&self.url)
+            .headers(self.headers.clone())
+            .query(&[("query", &self.query)])
+            .body(reqwest::Body::wrap_stream(stream))
+            .send()
+            .await?;
+
+        if res.status() != reqwest::StatusCode::OK {
+            anyhow::bail!("error writing to clickhouse: {}", res.text().await?);
+        }
```

Comment:
Should we implement a retry/exponential backoff logic before bailing out here?

Reply:
Yes, in the next iteration. First we have to test how this change performs in production